### PR TITLE
JS: Make the edges of API-graphs into IPA types

### PR DESCRIPTION
**Pull request by erik-krogh**
While looking at https://github.com/github/codeql/pull/6571 I kept running into performance issues related to API-graphs.  
The RA would often contain an `inverseappend[..]`, where it would do string-comparisons on every edge in order to extract the member-edges.  
It sometimes took a lot of effort to avoid that bad join. 

The obvious solution is to convert the edges into an IPA-type.  
This gives a more complicated implementation, but it's an implementation where future use of API-graphs should perform better.  
It also allows to `cache` the entire public API of API-graphs. 

I encountered a bunch of bad join order from this refactoring, so those were fixed along the way. 

--- 

[A multi-threaded evaluation](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/apiLabel2-nightly-security/reports) shows a limited performance improvement.  
[Running single-threaded](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/apiLabel2-nightly-security-single-threaded/reports) shows a more noticeable performance improvement. 

--- 

The `cache more predicates` and `optimizations in global data flow` commits have no impact on performance on their own.  
([Evaluation of just those two commits](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/apiLabel2JustOp__nightly__security-extend/reports)).   
And the `make ApiLabel into a IPA type` commit also has no impact on performance on its own ([evaluation](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/apiLabel2-just-api-nightly-security/reports)).

### CODE REVIEW

**Comment by asgerf** [comment]
> It also allows to cache the entire public API of API-graphs.

Have you tried this without an IPA type? We could have a predicate containing all labels occurring the graph, and use that to eliminate the binding sets and cache everything.

**Comment by hvitved** [comment]
I had also thought about making this change for Ruby, so perhaps I should give it a go.

**Comment by erik-krogh** [comment]
> > It also allows to cache the entire public API of API-graphs.
> 
> Have you tried this without an IPA type? We could have a predicate containing all labels occurring the graph, and use that to eliminate the binding sets and cache everything.

The bad join orders could still happen if we just eliminate the binding sets.  
Something like `result = someAPINode.getASuccessor(API::Label::member(someName))` could still introduce an `inverseAppend`.  

But I'll give it a go.

**Comment by erik-krogh** [comment]
> > It also allows to cache the entire public API of API-graphs.
> 
> Have you tried this without an IPA type? We could have a predicate containing all labels occurring the graph, and use that to eliminate the binding sets and cache everything.

I tried two experiments where I removed the IPA type.   
Both experiments showed neutral performance, so I still think we should try the IPA types. 

1: Just `cache` the public API ([implementation](https://github.com/erik-krogh/ql/commit/65f0c245218789ffed4fc77a2c9e25445ab4e0e1)). 
Evaluations: [multi-threaded](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/apiLabel2-just-cache-public-api-nightly-security/reports), and [single-threaded](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/apiLabel2-just-cache-single-thread/reports) both show about neutral performance (0.993 and 1.001 respectively). 

2: Same code as the current PR, but revert back to labels as strings ([implementation](https://github.com/erik-krogh/ql/commits/apiLabel2NoIPA)).    
Evaluations: [multi-threaded](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/apiLabel2-just-cache-public-nightly-security/reports), and [single-threaded](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/apiLabel2-noIPA-single-thread/reports) shows that there might be a tiny performance improvement (0.997 and 0.988 respectively).

**Review by asgerf — COMMENTED** [review]
Ok, just a few comments otherwise LGTM

File: javascript/ql/lib/semmle/javascript/ApiGraphs.qll [hunk 1]
```diff
@@ -996,89 +999,224 @@ module API {
...
   class NewNode extends InvokeNode, DataFlow::NewNode { }
-}
 
-private module Label {
-  /** Gets the edge label for the module `m`. */
-  bindingset[m]
-  bindingset[result]
-  string mod(string m) { result = "module " + m }
+  /** A label in the API-graph */
+  abstract class ApiLabel extends Label::TLabel {
```
[the remark this answers is not in the document]
Comment (erik-krogh) [reply javascript/ql/lib/semmle/javascript/ApiGraphs.qll]: No, there is not. Good point.  
I'll make the class non-abstract.

**Pull request merged**